tests/ClientProvider/ClientProvider.php

```diff
@@ -0,0 +1,212 @@
+    }
+
+    // CREATORS
+
+    /**
+     * @param array $config
+     * @param bool $useExistingBranch
+     * @return Client
+     */
+    public function createClientForCurrentTest($config = [], $useExistingBranch = false)
+    {
+        if ($this->testCase->usesDataProvider()) {
+            if ($this->testCase->getProvidedData()[0] === self::DEFAULT_BRANCH) {
+                return $this->getDefaultClient($config);
+            } elseif ($this->testCase->getProvidedData()[0] === self::DEV_BRANCH) {
+                return $this->getDevBranchClient($config, $useExistingBranch);
+            }
+        }
+
+        return $this->getDefaultClient();
+    }
+
+    /**
+     * @param array $config
+     * @param bool $useExistingBranch
+     * @return BranchAwareClient
+     */
+    public function createBranchAwareClientForCurrentTest($config = [], $useExistingBranch = false)
+    {
+        if ($this->testCase->usesDataProvider()) {
+            if ($this->testCase->getProvidedData()[0] === self::DEFAULT_BRANCH) {
+                return $this->getDefaultBranchClient($config);
+            } elseif ($this->testCase->getProvidedData()[0] === self::DEV_BRANCH) {
+                return $this->getDevBranchClient($config, $useExistingBranch);
+            }
+        }
+
+        return $this->getDefaultBranchClient();
+    }
+
+    /**
+     * @param array $config
+     * @param bool $useExistingBranch
+     * @return GuzzleClient
+     */
+    public function createGuzzleClientForCurrentTest($config, $useExistingBranch = false)
+    {
+        if ($this->testCase->usesDataProvider()) {
+            if ($this->testCase->getProvidedData()[0] === self::DEFAULT_BRANCH) {
+                return $this->getGuzzleClient($config);
+            } elseif ($this->testCase->getProvidedData()[0] === self::DEV_BRANCH) {
+                return $this->getBranchAwareGuzzleClient($config, $useExistingBranch);
+            }
+        }
+
+        return $this->getGuzzleClient($config);
+    }
+
+    // CLIENTS
+
+    /**
+     * @param array $config
+     * @return Client
+     */
+    public function getDefaultClient($config = [])
+    {
+        if ($config) {
+            return $this->testCase->getClient($config);
+        } else {
+            return $this->testCase->getDefaultClient();
+        }
+    }
+
+    /**
+     * @param array $config
+     * @return BranchAwareClient
+     */
+    public function getDefaultBranchClient($config = [])
+    {
+        $branchId = $this->testCase->getDefaultBranchId($this->testCase);
+        
+        if ($config) {
+            return $this->testCase->getBranchAwareClient($branchId, $config);
+        } else {
+            return $this->testCase->getBranchAwareDefaultClient($branchId);
+        }
+    }
+
+    /**
+     * @param array $config
+     * @param bool $useExistingBranch
+     * @return BranchAwareClient
+     */
+    public function getDevBranchClient($config = [], $useExistingBranch = false)
```

Comment:
tady se mi akorát furt nelíbí ten flag parametr, ale couž, je to detail :)

Reply:
Vůbec by mě neurazilo, kdyby ta metoda neřešila branch a prostě chtěla branchId. Co si pamatuju, tak to ničemu nevadí. Prostě Single Responsibility Principle - řeší to klienta, umí to řešit jen klienta. Chceš tam řešit branche, tak se nejdřív zeptej něčeho co řeší branche a sem dej jen ten výsledek :) 

V set up metodě by sis to explicitně zavolal "createBranchForTestCase" a poslal bys to sem. 
V optimálním scénáři "klient provider" providuje clienta a nezakládá branche. Od to by byl "BranchProvider" nebo tak něco - ten by ti zas mohl buď dát existující branch nebo založit novou. 

Ale zas to nechci zbytečně hrotit :)